STN-294 : Refactor External Links

heatherdesigns opened the pull request:
# [STN-294](https://sparkbox.atlassian.net/browse/STN-294) READY FOR REVIEW

## Summary
This PR fixes external link bugs and implements design tweaks to create consistency.

The external link styles in this PR have been reviewed by a designer.

:bug:
External link icon was missing hover styles on node detail pages.

:bug:
Within WYSIWYG main content, the external link icon was not positioned correctly when the link wrapped onto a second line.
_Example:_
<img width="818" alt="two_lines" src="https://user-images.githubusercontent.com/12678977/79123246-b7faec80-7d67-11ea-8a49-412a0be301f8.png">

:bug: + :art:
Update main content links to use underlines upon hover rather than border-bottom. This resolves a bug with the border-bottom extends beyond the text when the link text wraps unto a second line. It also creates design consistency across all links.
_Example:_
<img width="721" alt="update to underline" src="https://user-images.githubusercontent.com/12678977/79123411-1aec8380-7d68-11ea-8d0e-1e46fcc002c5.png">

🤖 
Refactor code base as needed to ensure external link styles are being applied consistently.

## This PR has been browser tested in
- [x] Safari
- [x] Chrome
- [x] Firefox (Mac & Windows)
- [x] Edge
- [x] IE
Firefox and IE have issues that need to be resolved. See comment below.

## Need Review By (Date)
asap

## Urgency
high

## Steps to Test
1. In the CLI, run `npm run test` to confirm tests pass.
2. Confirm the following bugs have been squashed:

- [ ] Go to the [News Node Detail](http://economics.suhumsci.loc/news/joeys-test-news-item) page. Hover over the external link text. Confirm the external link icon darkens when hovering over the link text. _Note:_ The fix implemented should cover all node detail pages. For example, go to a person node detail page such as http://economics.suhumsci.loc/people/person-all-fields-filled and confirm the external link icon darkens when hovering over the link text.

- [ ] Go to the [QA Real Econ Content](http://economics.suhumsci.loc/qa/qa-real-econ-content) page. Scroll through the main content. Find an external link which wraps onto a second line. `Stanford Institute for Economic Policy Research (SIEPR)` is a good link to look for. Verify the the icon sits to the right of the link.

- [ ] Go to [QA Page](http://economics.suhumsci.loc/qa/qa-page). Scroll down the page until you find a section of horizontal news cards. Adjust the page width until the external link wraps unto a second line. Confirm the underline resolves the issue previously created by border-bottom.

- [ ] Review the link styles in EVERYWHERE! Look and notate any inconsistencies.

## PR Checklist
- [PR Checklist](https://gist.github.com/sherakama/0ba17601381e3adbe0cad566ad4d80a5)
- [Sparkbox PR Checklist](../docs/SparkboxPRChecklist.md)

Commit:
feat: refactor external link icons

- add display property to link styles to allow external link icons to be positioned correctly when they are used in accompaniment with long text links which wrapped onto multiple lines

Commit:
fix: give external link icons on news node detail pages and news view list items a darkened hover state

Commit:
feat: add comments for next steps

Commit:
feat: refactor external link icons

- we have two mixins for external link icons. Refactor and eliminate one of the mixins.

Commit:
feat: implement design pairing feedback for hb-pills and external links

Commit:
feat: complete refactor using the new hb-ext-icon-color mixin

heatherdesigns commented on docroot/themes/humsci/humsci_basic/src/scss/tools/_mixins.text.scss:
```diff
@@ -167,14 +169,6 @@
-    background-repeat: no-repeat;
-    background-position: center;
-  }
-}
```
This has been renamed to `hb-ext-icon-color` and moved into the icon mixins.

heatherdesigns commented:
### Browser Testing Results
:bug: Firefox (Windows & Mac) and IE have issues with the external link icon when the url text wraps unto a second line. These screenshots are from the `Real Content Econ` page.

#### Firefox
<img width="921" alt="firefox real econ external link" src="https://user-images.githubusercontent.com/12678977/79173717-6897d900-7dc6-11ea-9edc-c5a920af9e5e.png">

#### IE
<img width="870" alt="IE" src="https://user-images.githubusercontent.com/12678977/79173721-6d5c8d00-7dc6-11ea-8d47-226bbb03b7eb.png">

Commit:
feat: fix Firefox positioning bug

Commit:
fix: color of external link icons on hover

silverli commented on docroot/themes/humsci/humsci_basic/src/scss/tools/_mixins.icons.scss:
```diff
@@ -27,45 +27,56 @@
 
-// Used with the decanter-external-link class to add an external link icon
+// Used alongside the hb-link--inline mixin when an external icon is needed.
+// And within the fa-ext class in block-layout.scss to add an external link icon.
```
I am still struggling to understand when this mixin is used vs the one below called `hb-ext-icon-color`. could you add a couple more comments about each of their use cases?

silverli commented on docroot/themes/humsci/humsci_basic/src/scss/utilities/_wysiwyg-text-area.scss:
```diff
@@ -255,7 +255,7 @@ blockquote:nth-child(n) {
 
   .hb-local-footer & {
-    margin: 0 0 2rem; // adjust blockquote browser spacing defaults when placed in the local footer
+    margin: 0 0 hb-calculate-rems(20px); // adjust blockquote browser spacing defaults when placed in the local footer
```
good catch!

Commit:
feat: clarify icon usage comments

silverli commented on docroot/themes/humsci/humsci_basic/src/scss/tools/_mixins.icons.scss:
```diff
@@ -27,7 +27,9 @@
-// Used alongside the hb-link--inline mixin when an external icon is needed.
+// The external link icon base implementation which sets the sizing and spacing.
+// This has two use cases.
+// It is used alongside the hb-link--inline mixin when an external icon is needed. The WYSIWYG hs-external-link class is an example of this.
```
great!! thanks for including this

silverli commented on docroot/themes/humsci/humsci_basic/src/scss/tools/_mixins.icons.scss:
```diff
@@ -69,6 +71,8 @@
 }
 
 // Return a specific color of the fa-ext icon
+// This mixin can be used to trigger the icon to change colors when the
```
nice

heatherdesigns commented on docroot/themes/humsci/humsci_basic/src/scss/utilities/_lists.scss:
```diff
@@ -76,7 +76,9 @@
-      padding-right: hb-calculate-rems(15px);
+      &::after {
+        background-color: transparent;
+      }
```
We may be able to remove this if we find another solution for the background-color of the icon.

Commit:
feat: external link icon background bug solving in progress

- Fix alignment of pills that have external links after the addition of the height to the external link icon.

Commit:
feat: refactor hb-link--inline mixin to use background-image rather than box-shadow

Commit:
feat: clean up

- remove comments from hb-link--inline mixin
- add IE fallbacks to hb-link--inline mixin
- replace instances of box-shadow when use with anchor tags that have icons with background-image

Commit:
feat: adjust base anchor link selectors for external links

silverli commented on docroot/themes/humsci/humsci_basic/src/scss/elements/_base.scss:
```diff
@@ -46,6 +46,10 @@ a:not([class]) {
   @include hb-link--inline;
 }
 
+a[href*="//"][data-extlink] {
```
can you include a comment about this strategy?

silverli commented on docroot/themes/humsci/humsci_basic/src/scss/tools/_mixins.text.scss:
```diff
@@ -133,23 +134,43 @@
+    background-position-y: -100%;
+    background-repeat: no-repeat;
+
+    // For links without an icon
```
can you also write a comment about what this does for external links specifically? thanks!

Commit:
feat: implement review feedback

- add comments to explain new functionality of hb-link--inline mixin
- add table cell support for external link icon hover state

silverli commented on docroot/themes/humsci/humsci_basic/src/scss/tools/_mixins.text.scss:
```diff
@@ -215,6 +215,8 @@
 
     // For links with an icon
+    // Changes the value passed into background-image to
+    // create space for the external link space to exist
```
🎉

silverli reviewed (approved):
Looks amazing. Thanks for all your hard work on this.
